l5kit/l5kit/environment/envs/l5_env.py

```diff
@@ -0,0 +1,333 @@
+            ego_output = self._convert_action_to_ego_output(action)
+            self.ego_output_dict = ego_output
+
+            if self.cle:
+                # In closed loop training, the raster is updated according to predicted ego positions.
+                self.simulator.update_ego(self.sim_dataset, next_frame_index, self.ego_input_dict, self.ego_output_dict)
+
+            ego_frame_in_out = self.simulator.get_ego_in_out(self.ego_input_dict, self.ego_output_dict,
+                                                             self.simulator.keys_to_exclude)
+            self.ego_ins_outs[self.scene_index].append(ego_frame_in_out[self.scene_index])
+
+        # generate simulated_outputs
+        simulated_outputs = SimulationOutputCLE(self.scene_index, self.sim_dataset, self.ego_ins_outs,
+                                                self.agents_ins_outs)
+
+        # reward calculation
+        reward = self.reward.get_reward(self.frame_index, [simulated_outputs])
+
+        # done is True when episode ends
+        done = episode_over
+
+        # Optionally we can pass additional info
+        # We are using "info" to output rewards and simulated outputs (during evaluation)
+        info: Dict[str, Any]
+        info = {'reward_tot': reward["total"], 'reward_dist': reward["distance"], 'reward_yaw': reward["yaw"]}
+        if done and self.return_info:
+            info = {"sim_outs": self.get_episode_outputs(), "reward_tot": reward["total"],
+                    "reward_dist": reward["distance"], "reward_yaw": reward["yaw"]}
+
+        # Get next obs
+        self.frame_index += 1
+        obs = self._get_obs(self.frame_index, episode_over)
+
+        # return obs, reward, done, info
+        return GymStepOutput(obs, reward["total"], done, info)
+
+    def get_episode_outputs(self) -> List[EpisodeOutputGym]:
+        """Generate and return the outputs at the end of the episode.
+
+        :return: List of episode outputs
+        """
+        episode_outputs = [EpisodeOutputGym(self.scene_index, self.sim_dataset, self.ego_ins_outs,
+                                            self.agents_ins_outs)]
+        return episode_outputs
+
+    def render(self) -> None:
+        """Render a frame during the simulation
+        """
+        raise NotImplementedError
+
+    def _get_obs(self, frame_index: int, episode_over: bool) -> Dict[str, np.ndarray]:
+        """Get the observation corresponding to a given frame index in the scene.
+
+        :param frame_index: the index of the frame which provides the observation
+        :param episode_over: flag to determine if the episode is over
+        :return: the observation corresponding to the frame index
+        """
+        if episode_over:
+            frame_index = 0  # Dummy final obs (when episode_over)
+
+        ego_input = self.sim_dataset.rasterise_frame_batch(frame_index)
+        self.ego_input_dict = convert_to_numpy(ego_input[0])
```

Comment:
Same comment here, the `rasterise_frame_batch` return type is `List[Dict[str, np.ndarray]]`, so it seems to be already a numpy type correct ?